src/box/box.cc

```diff
@@ -3106,6 +3106,7 @@ box_demote(void)
 		if (raft->state != RAFT_STATE_LEADER)
 			return 0;
 	}
+	raft_update_last_demoted_term(raft);
```

Comment:
TBH I don't really like that we introduce new concepts to the raft library. Maybe we can reuse some existing functions here? It seems, you may replace `box_trigger_elections()` with `box_raft_leader_step_off()`: this way the leader will drop its leadership, but will still watch out for other candidates, and will re-start elections if noone else is found.

Yes, this way the demoted node may become leader right in the next term, but this will only happen if there are no other candidates.

@Serpentian, what do you think of this?

Reply:
@sergepetrenko `box_raft_leader_step_off` indeed resigns from the leader state, but right after that it calls `raft_restore`, which schedules vote for itself, so we won't really fix the issue. Just tested this, it's still diff

Initially Georgiy had version with `raft_cfg_is_candidate_later`:

```c
	bool is_cfg_candidate_saved = raft->is_cfg_candidate;
	/* This node must not be a candidate in the current elections. */
	raft_cfg_is_candidate_later(raft, false);
	int rc = box_trigger_elections();
	/*
	 * Even if we fail here, we definitely started a new term, so it is
	 * preferable that we are not a candidate in the current term.
	 */
	raft_cfg_is_candidate_later(raft, is_cfg_candidate_saved);
	if (rc != 0)
		return rc;
```

It did prevent node from voting, but it also prevented it from starting the new term at all, which is not great. See [this comment](https://github.com/tarantool/tarantool/pull/10003#discussion_r1597647241). Maybe we can return to this implementation and modify it somehow. Do you like it better? If so, we need to figure out a way, to modify it in the way, that it properly starts new term after resigning without modifying `lib/raft`.

Reply:
>  but right after that it calls raft_restore, which schedules vote for itself, so we won't really fix the issue.

But why? I don't see why this would happen by looking at the code: https://github.com/tarantool/tarantool/blob/39af9fbea6a9c13fe438b9546856416f7bdf1917/src/lib/raft/raft.c#L1139-L1154

Shouldn't we end up in `raft_sm_wait_leader_found` letting other candidates promote themselves?

My suggested diff is (note we skip `box_trigger_elections()` completely for `election_mode 'candidate'/'manual'`:
```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 447237f8da..c14613a4a0 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -3105,11 +3105,11 @@ box_demote(void)
                        return 0;
                if (raft->state != RAFT_STATE_LEADER)
                        return 0;
+               box_raft_leader_step_off();
+               return 0;
        }
        if (box_trigger_elections() != 0)
                return -1;
-       if (box_election_mode != ELECTION_MODE_OFF)
-               return 0;
        if (box_try_wait_confirm(2 * replication_synchro_timeout) < 0)
                return -1;
        int64_t wait_lsn = box_wait_limbo_acked(replication_synchro_timeout);
diff --git a/src/box/raft.c b/src/box/raft.c
index b5b97efe00..23eed9419e 100644
--- a/src/box/raft.c
+++ b/src/box/raft.c
@@ -308,7 +308,7 @@ box_raft_fence(void)
  * begin. After the death-timeout expires, it starts a new
  * round of elections.
  */
-static void
+void
 box_raft_leader_step_off(void)
 {
        struct raft *raft = box_raft();
diff --git a/src/box/raft.h b/src/box/raft.h
index 41611d27fc..6d3df7a567 100644
--- a/src/box/raft.h
+++ b/src/box/raft.h
@@ -144,6 +144,9 @@ box_raft_wait_term_persisted(void);
 void
 box_raft_set_election_fencing_mode(enum election_fencing_mode mode);
 
+void
+box_raft_leader_step_off(void);
+
 /**
  * Pause fencing. Instance will not resign its leader role when it looses
  * quorum even if fencing is enabled. Useful to prevent fencing while adding
```

Reply:
> But why?

Seems, I tested incorrect version. The diff, you propose indeed seems to work. But let's also refactor `box_demote` while we're here, I spent 10 minutes to figure out, what is going on there. Here's the proposed refactoring:

<details>

<summary>Diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 447237f8da..224a7de67f 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -3080,24 +3080,7 @@ box_demote(void)
 		return 0;
 
 	const struct raft *raft = box_raft();
-	if (box_election_mode == ELECTION_MODE_OFF) {
-		assert(raft->state == RAFT_STATE_FOLLOWER);
-		if (raft->leader != REPLICA_ID_NIL) {
-			diag_set(ClientError, ER_NOT_LEADER, raft->leader);
-			return -1;
-		}
-		if (txn_limbo.owner_id == REPLICA_ID_NIL)
-			return 0;
-		/*
-		 * If the limbo term is up to date with Raft, then it might have
-		 * a valid owner right now. Demotion would disrupt it. In this
-		 * case the user has to explicitly overthrow the old owner with
-		 * local promote(), or call demote() on the actual owner.
-		 */
-		if (txn_limbo.promote_greatest_term == raft->term &&
-		    txn_limbo.owner_id != instance_id)
-			return 0;
-	} else {
+	if (box_election_mode != ELECTION_MODE_OFF) {
 		if (txn_limbo_replica_term(&txn_limbo, instance_id) !=
 		    raft->term)
 			return 0;
@@ -3105,11 +3088,28 @@ box_demote(void)
 			return 0;
 		if (raft->state != RAFT_STATE_LEADER)
 			return 0;
+		box_raft_leader_step_off();
+		return 0;
 	}
-	if (box_trigger_elections() != 0)
+
+	assert(raft->state == RAFT_STATE_FOLLOWER);
+	if (raft->leader != REPLICA_ID_NIL) {
+		diag_set(ClientError, ER_NOT_LEADER, raft->leader);
 		return -1;
-	if (box_election_mode != ELECTION_MODE_OFF)
+	}
+	if (txn_limbo.owner_id == REPLICA_ID_NIL)
 		return 0;
+	/*
+	 * If the limbo term is up to date with Raft, then it might have
+	 * a valid owner right now. Demotion would disrupt it. In this
+	 * case the user has to explicitly overthrow the old owner with
+	 * local promote(), or call demote() on the actual owner.
+	 */
+	if (txn_limbo.promote_greatest_term == raft->term &&
+			txn_limbo.owner_id != instance_id)
+		return 0;
+	if (box_trigger_elections() != 0)
+		return -1;
 	if (box_try_wait_confirm(2 * replication_synchro_timeout) < 0)
 		return -1;
 	int64_t wait_lsn = box_wait_limbo_acked(replication_synchro_timeout);
diff --git a/src/box/raft.c b/src/box/raft.c
index b5b97efe00..23eed9419e 100644
--- a/src/box/raft.c
+++ b/src/box/raft.c
@@ -308,7 +308,7 @@ box_raft_fence(void)
  * begin. After the death-timeout expires, it starts a new
  * round of elections.
  */
-static void
+void
 box_raft_leader_step_off(void)
 {
 	struct raft *raft = box_raft();
diff --git a/src/box/raft.h b/src/box/raft.h
index 41611d27fc..3b43d6fa38 100644
--- a/src/box/raft.h
+++ b/src/box/raft.h
@@ -158,6 +158,9 @@ box_raft_init(void);
 void
 box_raft_free(void);
 
+void
+box_raft_leader_step_off(void);
+
 #if defined(__cplusplus)
 }
 #endif

```

</details>

Reply:
Nice!

Reply:
@Serpentian applied your refactoring as a separate patch, @sergepetrenko applied your solution, thanks for the feedback guys!

Reply:
It seems like the current solution creates a regression to 1afe2274: in manual election mode, the demoted leader would always bump the term (now it doesn't), which allowed other nodes to disown the synchronous queue when they issued `box.ctl.demote` with `box.cfg{election_mode = 'off'`:
https://github.com/tarantool/tarantool/blob/8ed6104987bf54d82a7a79cfa2726a57dd889894/src/box/box.cc#L3091-L3099

It breaks this test:
https://github.com/tarantool/tarantool/blob/8ed6104987bf54d82a7a79cfa2726a57dd889894/test/replication-luatest/gh_6860_election_off_demote_test.lua#L101-L131

Reply:
> By the way, while we are here, why do we need both of these checks?
> 
> https://github.com/tarantool/tarantool/blob/ef3152cdc249045be8250c65d47b1a002e49467e/src/box/box.cc#L3104-L3107
> 
> Doesn't the leader always own the synchronous queue?

The desired state for a leader is to own the synchronous queue. But this doesn't happen instantly, just because a node is promoted leader.

For example, when just being elected, Raft leader starts the synchronous queue transition, but it happens asynchronously. See `box_raft_update_synchro_queue()`. So there's a point in time, when `election_state == 'leader'` but `synchro_queue_owner` is someone else.

Also, when a leader steps off or is being overthrown, its state is `follower`, but it still is the synchro queue owner.

Reply:
> TBH, when I was looking for a solution, this snippet boggled me:
> 
> https://github.com/tarantool/tarantool/blob/39af9fbea6a9c13fe438b9546856416f7bdf1917/src/lib/raft/raft.c#L1139-L1154
> 
> It's not immediately obvious that the leader's `leader_witness_map` is always zero, and it's not noted in the comment to `leader_witness_map`.

I re-checked the code, and indeed it seems the leader doesn't keep track of `leader_witness_map`. And nither does it run the election timer (or leader death timer). Which seems quite logical.

But I don't get what's wrong with the snippet you provide. Could you elaborate, please? We know that the node can't be a current leader, because this is a `raft_start_candidate` call (meaning that a node now may become candidate. It doesn't mean it enters "raft candidate" state immediately):
https://github.com/tarantool/tarantool/blob/39af9fbea6a9c13fe438b9546856416f7bdf1917/src/lib/raft/raft.c#L1131-L1139

Reply:
> But I don't get what's wrong with the snippet you provide. Could you elaborate, please? 

Nothing wrong, just noting that it wasn't obvious for me that this branch would be taken by a just demoted leader, and I didn't take into account — would be nice to drop a comment near `leader_witness_map` that it's always zero for the leader.

Reply:
> To fix this, I suggest to leave everything as is for manual election mode, i.e., call `box_trigger_elections` instead of `box_raft_leader_step_off`. It's obvious that #9855 doesn't occur for manual elections, however, the fact that the demoted leader bumps the Raft term is essential for allowing `box.ctl.demote` to work properly on nodes with disabled elections.
> 
> The problem above can technically still happen with a `voter` node in automatic elections, if nobody triggers new elections, and some `voter` node gets its elections disabled, but I guess it's not a legit case we should try to fix.

I think this is a case where you better rewrite the test than alter `demote()` behavior.
I like the idea that `demote()` behaves consistently with "elections enabled" (`election_mode='manual'/'candidate'). We better keep it this way. I mean, it always keeps the same term and it never writes an actual DEMOTE request.

The problem you mention (with one candidate and many voters) is the same as if all nodes were manual and the leader died. Noone could demote the queue without first promoting itself.

And if the leader is just fine, there's no point to 'demote' it in election_mode = 'manual'. You may first set its election_mode to off and only then issue a demote. Or, alternatively, you may demote it twice. Before and after turning off elections.

The ticket which got fixed by https://github.com/tarantool/tarantool/commit/1afe2274d9155fc19d228237d04d1a1ddfa17270 complained about another issue: even the former leader couldn't release synchro queue without first calling box.ctl.promote().

Reply:
I don't understand how to create a situation where the synchro queue term is behind the election term with the current patch (which is what the test relies on).

Reply:
For example, you may increase synchro quorum and make a node try to promote. It won't succeed. But the term will be bumped. And the other node will be able to demote.

Although it seems the test becomes almost pointless after the demote rewrite.

Reply:
> But the promote still succeeds (AFAIC, because the promote is currently not synchronous?). Or what am I doing wrong?

Oh, yes, I meant promote with `election_mode = 'manual'`. It would trigger elections which the node couldn't win due to huge quorum.
With `election_mode = 'off'` indeed the user is free to do whatever he wants to and may promote an instance without a quorum.

Reply:
I also thought about it, but I got the same result:

```console
tarantool> os.execute('rm -rf *.snap *.xlog *.vylog 512')
---
- 0
...

tarantool> box.cfg{replication_synchro_quorum = 4}
2024-05-27 12:58:50.063 [13625] main/104/interactive I> Tarantool 3.1.0-entrypoint-238-gb3284e10c0 Darwin-x86_64-Debug
2024-05-27 12:58:50.064 [13625] main/104/interactive I> log level 5 (INFO)
2024-05-27 12:58:50.064 [13625] main/104/interactive I> wal/engine cleanup is paused
2024-05-27 12:58:50.065 [13625] main/104/interactive I> mapping 268435456 bytes for memtx tuple arena...
2024-05-27 12:58:50.065 [13625] main/104/interactive I> Actual slab_alloc_factor calculated on the basis of desired slab_alloc_factor = 1.044274
2024-05-27 12:58:50.065 [13625] main/104/interactive I> mapping 134217728 bytes for vinyl tuple arena...
2024-05-27 12:58:50.077 [13625] main/104/interactive I> RAFT: fencing paused
2024-05-27 12:58:50.077 [13625] main/104/interactive I> instance uuid 24e89994-b751-4aae-99a0-78065d23ecb9
2024-05-27 12:58:50.078 [13625] main/104/interactive I> initializing an empty data directory
2024-05-27 12:58:50.133 [13625] main/104/interactive I> assigned id 1 to replica 24e89994-b751-4aae-99a0-78065d23ecb9
2024-05-27 12:58:50.134 [13625] main/104/interactive I> replicaset uuid 737f2a1d-99f9-4d0f-88af-18e460d83edc
2024-05-27 12:58:50.137 [13625] snapshot/101/main I> saving snapshot `./00000000000000000000.snap.inprogress'
2024-05-27 12:58:50.141 [13625] snapshot/101/main I> done
2024-05-27 12:58:50.143 [13625] main/104/interactive I> ready to accept requests
2024-05-27 12:58:50.143 [13625] main/105/gc I> wal/engine cleanup is resumed
2024-05-27 12:58:50.143 [13625] main/104/interactive/box.load_cfg I> set 'replication_synchro_quorum' configuration option to 4
2024-05-27 12:58:50.143 [13625] main/106/checkpoint_daemon I> scheduled next checkpoint for Mon May 27 13:59:34 2024
---
...

tarantool> box.cfg{election_mode = 'manual'}
2024-05-27 12:58:58.366 [13625] main/104/interactive I> RAFT: start state machine
2024-05-27 12:58:58.366 [13625] main/104/interactive/box.load_cfg I> set 'election_mode' configuration option to "manual"
---
...

tarantool> box.ctl.promote()
2024-05-27 12:59:04.798 [13625] main/104/interactive I> RAFT: bump term to 2, follow
2024-05-27 12:59:04.798 [13625] main/104/interactive I> RAFT: vote for 1, follow
2024-05-27 12:59:04.799 [13625] main/114/raft_worker I> RAFT: persisted state {term: 2, vote: 1}
2024-05-27 12:59:04.799 [13625] main/114/raft_worker I> RAFT: enter leader state with quorum 1
---
...

```

Reply:
Actually, it's enough to just stop one of the two servers. 

Fixed: stopped one of the servers to make box.ctl.promote fail and then restarted it.

Reply:
Thanks for the fixes!